libs/brig-types/src/Brig/Types/Common.hs

```diff
@@ -150,6 +150,45 @@ instance FromJSON PhoneBudgetTimeout where
 instance ToJSON PhoneBudgetTimeout where
     toJSON (PhoneBudgetTimeout t) = object [ "expires_in" .= t ]
 
+-----------------------------------------------------------------------------
+-- PhonePrefix (for excluding from SMS/calling)
+
+newtype PhonePrefix = PhonePrefix { fromPhonePrefix :: Text } deriving (Eq, Show, ToJSON)
+
+-- | Parses a phone number prefix with a mandatory leading '+'.
+parsePhonePrefix :: Text -> Maybe PhonePrefix
+parsePhonePrefix p
+    | isValidPhonePrefix p = Just $! PhonePrefix p
+    | otherwise      = Nothing
+
+-- | Checks whether a phone number prefix is valid,
+-- i.e. it is like a E.164 format phone number, but shorter
+-- (with a mandatory leading '+', followed by 1-15 digits.)
+isValidPhonePrefix :: Text -> Bool
+isValidPhonePrefix = either (const False) (const True) . parseOnly e164Prefix
+  where
+    e164Prefix = char '+' *> count 1 digit *> count 14 (optional digit) *> endOfInput
+
+phonePrefixes :: Phone -> [PhonePrefix]
+phonePrefixes p = allPrefixes (fromPhone p)
+
+-- | get all valid prefixes of a phone number or phone number prefix
+-- e.g. from +123456789 get prefixes ["+1", "+12", "+123", ..., "+123456789" ]
+allPrefixes :: Text -> [PhonePrefix]
+allPrefixes t = catMaybes $ parsePhonePrefix <$> Text.inits t
+
+instance FromJSON PhonePrefix where
+    parseJSON (String s) = case parsePhonePrefix s of
```

Comment:
`withText` is better than `mempty` because it gives a descriptive error message in case of JSON type mismatch.

Reply:
thanks.